File: 2501-longest-square-streak-in-an-array/2501-longest-square-streak-in-an-array.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var longestSquareStreak = function(nums) {
    let map = new Map();
    let mlen = -1;
    
    nums.sort((a,b) => a - b);
    
    for(let num of nums){
        map.set(num, map.get(num)+1 || 1);
    }
    
    const sq = (num) =>{
        let ans = 1;
        num *= num;
        while(map.get(num)){
            ans++;
            num *= num;
        }
        return ans == 1? -1: ans;
    }
    
    for(let num of nums){
        mlen = Math.max(sq(num), mlen)
    }
    return mlen
};

File: 2501-longest-square-streak-in-an-array/2501-longest-square-streak-in-an-array.java
import java.util.*;

class Solution {
    public int longestSquareStreak(int[] nums) {
        int maxLen = 0;

        Set<Integer> set = new HashSet<>();
        for (int n : nums) {
            set.add(n);
        }

        Set<Integer> visited = new HashSet<>();
        
        for (int n : nums) {
            if (visited.contains(n)) continue;
            int len = 1;
            int current = n;

            while (current > 0 && current <= Math.sqrt(Integer.MAX_VALUE) && set.contains(current * current)) {
                visited.add(current);
                current *= current;
                len++;
            }
            maxLen = Math.max(maxLen, len);
        }

        return maxLen > 1? maxLen : -1;
    }
}